### PR TITLE
Add profiling hooks on systhreads

### PR DESCRIPTION
**Pull request by lefessan**
This PR adds hooks on systhreads to be able to monitor their behavior. The hooks are defined in the standard runtime, to be always available, but only used when system threads are being used. 

There are 2 kinds of hooks:
- the first hook is called when scanning threads roots, providing the thread being scanned before its local roots are scanned
- the second hook is called everytime a thread change happens, i.e. when a thread is created, is scheduled or yield.

### CODE REVIEW

File: otherlibs/systhreads/st_stubs.c [hunk 1]
```diff
@@ -159,6 +165,7 @@ static void caml_thread_scan_roots(scanning_action action)
 static inline void caml_thread_save_runtime_state(void)
 {
 #ifdef NATIVE_CODE
+  curr_thread->top_of_stack = caml_top_of_stack;
```
Comment (klakplok) [inline otherlibs/systhreads/st_stubs.c]: Is this change related to the hooks ?

File: otherlibs/systhreads/st_stubs.c [hunk 2]
```diff
@@ -291,7 +301,9 @@ static uintnat caml_thread_stack_usage(void)
        th != curr_thread;
        th = th->next) {
 #ifdef NATIVE_CODE
-    sz += (value *) th->top_of_stack - (value *) th->bottom_of_stack;
+    if(th->top_of_stack != NULL && th->bottom_of_stack != NULL &&
+       th->top_of_stack > th->bottom_of_stack)
+      sz += (value *) th->top_of_stack - (value *) th->bottom_of_stack;
```
Comment (klakplok) [inline otherlibs/systhreads/st_stubs.c]: Is this change related to the hooks ? (3)

File: otherlibs/systhreads/st_stubs.c [hunk 3]
```diff
@@ -176,11 +183,14 @@ static inline void caml_thread_save_runtime_state(void)
...
   curr_thread->backtrace_buffer = backtrace_buffer;
   curr_thread->backtrace_last_exn = backtrace_last_exn;
+  ST_CHANGE_HOOK(curr_thread->uniq_ident,ST_HOOK_YIELD);
 }
 
 static inline void caml_thread_restore_runtime_state(void)
 {
+  ST_CHANGE_HOOK(curr_thread->uniq_ident,ST_HOOK_SCHEDULE);
 #ifdef NATIVE_CODE
+  caml_top_of_stack = curr_thread->top_of_stack;
```
Comment (klakplok) [inline otherlibs/systhreads/st_stubs.c]: Is this change related to the hooks ? (2)

File: byterun/caml/roots.h [hunk 4]
```diff
@@ -35,10 +35,48 @@ CAMLextern void caml_do_local_roots (scanning_action, value *, value *,
...
 CAMLextern void (*caml_scan_roots_hook) (scanning_action);
 
+
+/* Hooks to profile systhreads. */
+
+#define ST_NEXT_HOOK(thread_id,bottom,retaddr) \
+  if( st_next_hook != NULL ) st_next_hook(thread_id,bottom,retaddr)
+CAMLextern void (*st_next_hook)(value thread_id, char* bottom,
+                                        uintnat retaddr);
+/*
```
Comment (klakplok) [inline byterun/caml/roots.h]: Even if that is the intended purpose of this patch, I don't think the descriptions of the hooks should mention allocation profiling. You may simply say 'run OCaml code' instead of 'allocate'.
Also, I would replace `corresponding` by `The`.

File: otherlibs/systhreads/st_stubs.c [hunk 5]
```diff
@@ -86,6 +86,7 @@ struct caml_thread_struct {
   int backtrace_pos;            /* Saved backtrace_pos */
   backtrace_slot * backtrace_buffer;    /* Saved backtrace_buffer */
   value backtrace_last_exn;     /* Saved backtrace_last_exn (root) */
+  value uniq_ident;             /* a pre-copy of caml_thread_descr uid */
```
Comment (klakplok) [inline otherlibs/systhreads/st_stubs.c:100]: Maybe add `(immediate value)` to the comment.

File: byterun/caml/roots.h [hunk 6]
```diff
@@ -35,10 +35,48 @@ CAMLextern void caml_do_local_roots (scanning_action, value *, value *,
...
                                     struct caml__roots_block * local_roots);
-#endif
+#endif /* NATIVE_CODE */
 
 CAMLextern void (*caml_scan_roots_hook) (scanning_action);
 
+
+/* Hooks to profile systhreads. */
+
+#define ST_NEXT_HOOK(thread_id,bottom,retaddr) \
```
Comment (klakplok) [inline byterun/caml/roots.h]: I think that both macros should have a short description, not only the enumeration of hooks.

**Comment by klakplok** [comment]
I believe the change to be sound.
All hook calls seem to be in the right places.
It should not change the semantics when no hooks are defined.
I just made a few, mostly cosmetic remarks.

File: byterun/caml/roots.h [hunk 7]
```diff
@@ -35,10 +35,48 @@ CAMLextern void caml_do_local_roots (scanning_action, value *, value *,
 CAMLextern void caml_do_local_roots(scanning_action f, char * bottom_of_stack,
                                     uintnat last_retaddr, value * gc_regs,
                                     struct caml__roots_block * local_roots);
-#endif
+#endif /* NATIVE_CODE */
```
Comment (damiendoligez) [inline byterun/caml/roots.h:38]: If you want to put this annotation on `#endif` you should avoid confusion: invert the condition and exchange the `then` and `else` parts.

File: byterun/caml/roots.h [hunk 8]
```diff
@@ -35,10 +35,48 @@ CAMLextern void caml_do_local_roots (scanning_action, value *, value *,
...
 
 CAMLextern void (*caml_scan_roots_hook) (scanning_action);
 
+
```
Comment (damiendoligez) [inline byterun/caml/roots.h]: Same remark as for #696 : this has nothing to do with roots so it doesn't belong in this file.

File: byterun/caml/roots.h [hunk 9]
```diff
@@ -35,10 +35,48 @@ CAMLextern void caml_do_local_roots (scanning_action, value *, value *,
...
+/*
+ ST_HOOK_INIT: corresponding thread is the first thread of the
+    system. All preceeding allocations were done by it !
+ ST_HOOK_YIELD: corresponding thread is not going to allocate
+    anymore until scheduled again.
+ ST_HOOK_SCHEDULE: corresponding thread is being scheduled.
+    Further allocations are done by it !
+ ST_HOOK_STOP: corresponding thread is being terminated. It will
+    NEVER allocate again.
+ ST_HOOK_REINIT: corresponding thread is the first thread after a fork.
```
Comment (damiendoligez) [inline byterun/caml/roots.h]: I don't care what the numbers are, you must describe `INIT` and `REINIT` right next to each other.

File: byterun/caml/roots.h (hunk 6 above)
Comment (damiendoligez) [inline byterun/caml/roots.h]: The name of the macro (and the hook) is really weird. Since it's about scanning roots, it should be called something like `ST_ROOT_SCAN_HOOK` or something.

File: otherlibs/systhreads/st_stubs.c [hunk 10]
```diff
@@ -423,7 +442,9 @@ CAMLprim value caml_thread_initialize(value unit)   /* ML */
   /* Set up a thread info block for the current thread */
   curr_thread =
     (caml_thread_t) caml_stat_alloc(sizeof(struct caml_thread_struct));
-  curr_thread->descr = caml_thread_new_descriptor(Val_unit);
+  curr_thread->uniq_ident = caml_thread_new_ident();
+  ST_CHANGE_HOOK(curr_thread->uniq_ident,ST_HOOK_INIT);
```
Comment (damiendoligez) [inline otherlibs/systhreads/st_stubs.c]: Calling the hook at this point means it cannot go through `all_threads` and recover the thread's descriptor. Wouldn't it be better to call the hook after all the fields are initialized and the record is inserted in the list?

File: otherlibs/systhreads/st_stubs.c [hunk 11]
```diff
@@ -336,7 +348,14 @@ static caml_thread_t caml_thread_new_info(void)
...
 
-static value caml_thread_new_descriptor(value clos)
+static value caml_thread_new_ident()
+{
+   value id = Val_long(thread_next_ident);
+   thread_next_ident++;
+   return id;
+}
+
+static value caml_thread_new_descriptor(value clos, caml_thread_t th)
```
Comment (damiendoligez) [inline otherlibs/systhreads/st_stubs.c]: Why pass the (partially uninitialized) `caml_thread_t` instead of just the `uniq_ident` that you need? 
This is bad for maintainability.

File: otherlibs/systhreads/st_stubs.c [hunk 12]
```diff
@@ -572,6 +596,9 @@ CAMLexport int caml_c_thread_register(void)
 #endif
   /* Take master lock to protect access to the chaining of threads */
   st_masterlock_acquire(&caml_master_lock);
+  th->uniq_ident = caml_thread_new_ident();
```
Comment (damiendoligez) [inline otherlibs/systhreads/st_stubs.c:638]: You're setting `th->uniq_ident` without setting `Ident(th->descr)`. Aren't they supposed to be kept in sync?

[edit] The `descr` is set at line 620. Is there any way to pull these two closer? I'm worried about calling the hooks while the data structure is inconsistent.

**Comment by lefessan** [comment]
A new version:
- contains the commit from natdynlink_hook, that creates `caml/hooks.h`
- identifiers and macros renamed in `caml_*`
- `caml_thread_new_descriptor` now takes the identifier and not the partially initialized structure
- better documentation

Unfortunately, I don't think it is possible to move the hooks so that thread structures are always completely initialized when they are called, idem for the list of threads.

About `top_of_stack`, yes, it is unrelated to the hooks, but it fixes the computation of stack usage, I kept both in the same patch as they are both used by ocp-memprof...

**Comment by lefessan** [comment]
Maybe I should create `byterun/hooks.c` too, and move the hook definitions there ?

**Comment by damiendoligez** [comment]
> Maybe I should create `byterun/hooks.c` too, and move the hook definitions there ?

No, I think each hook definition is better located next to the code that calls it.

**Comment by damiendoligez** [comment]
> About `top_of_stack`, yes, it is unrelated to the hooks, but it fixes the computation of stack usage, I kept both in the same patch as they are both used by ocp-memprof...

Can you get a review of this point by a thread expert? Unlike the rest of the patch, it changes the behaviour of the system even when hooks are not used.

**Comment by lefessan** [comment]
> Can you get a review of this point by a thread expert? Unlike the rest of the patch, it changes the behaviour of the system even when hooks are not used.

Not really. `top_of_stack` is only used for the computation of stack space usage, so it cannot generate problems for people who don't use that function (and I don't think anybody uses that function, except memory profiling/debugging tools). The fix is quite obvious, as `top_of_stack` is not preserved when saving/restoring thread contexts, the computation of stack usage was false most of the time before the fix.

**Comment by lefessan** [comment]
@damiendoligez 
Actually, I am looking at the current implementation of backtraces, and `caml_get_current_callstack` in `asmrun/backtrace_prim.c` does use `caml_top_of_stack`. It's likely that, when systhreads are used, calling `caml_get_current_callstack` will fail since it will iter from the bottom of one thread's stack to the top of stack of the first thread of the application. So, this PR is actually fixing that bug, together with adding the hooks...

**Comment by lefessan** [comment]
I updated the PR to remove the fix of `caml_top_of_stack`, now moved to #996.

Except for the new management of thread idents that was reviewed by @damiendoligez (see comments above), I think it is now easy to see that the addition of these hooks is trivially harmless.

**Comment by damiendoligez** [comment]
@lefessan are you still interested in rebasing this PR at this point?

**Comment by gasche** [comment]
This PR is dead since at least 2018 with no sign of activity. Closing. We can always reopen if the PR gets rebased and people show more interest.
